### PR TITLE
Websocket continuation frames dropped if there are leftovers in the buffer from the previous read

### PR DESCRIPTION
**Pull request by pyrist**
Fix for #2928

### CODE REVIEW

**Comment by vinniefalco** [comment]
Is this a real bug?

**Comment by pyrist** [comment]
> Is this a real bug?

I believe so. It wouldn't make sense to me that the read loop should end after reading part of a header frame when the message and frame aren't finished.

**Comment by ashtum** [comment]
@pyrist, I couldn't reproduce the situation you described with the following code, which tests various message sizes and different points for chopping the input bytes.
```C++
for(std::size_t msg_size = 1 ; msg_size <= 10000; msg_size += 7)
{
    for(std::size_t i = 0 ;; ++i)
    {
        net::io_context ioc;
        stream<test::stream> wsc{ioc};
        stream<test::stream> wss{ioc};
        wsc.next_layer().connect(wss.next_layer());
        wsc.async_handshake(
            "localhost", "/", [](error_code){});
        wss.async_accept([](error_code){});
        ioc.run();
        ioc.restart();
        BEAST_EXPECT(wsc.is_open());
        BEAST_EXPECT(wss.is_open());
        auto sent_msg = std::string(msg_size, 'a');
        wss.write(net::buffer(sent_msg));

        if(i > wsc.next_layer().str().size())
            break; // we have tested all possible chunk sizes

        // steal the input buffer in order to write it in two steps
        std::string bytes = wsc.next_layer().str();
        wsc.next_layer().buffer().clear();

        // first part
        wss.next_layer().write_some(
            net::buffer(bytes.substr(0, i)));

        flat_buffer received_msg;
        wsc.async_read(
            received_msg,
            [](error_code, std::size_t){});

        ioc.poll();

        // second part
        wss.next_layer().write_some(
            net::buffer(bytes.substr(i)));

        ioc.run();
        BEAST_EXPECT(received_msg.size() == msg_size);
    }
}
```

**Comment by pyrist** [comment]
@ashtum Thanks for looking at it. I will look tomorrow to compare between what your code is doing and what is going wrong when I use it and see if I can make something reproducible.

File: include/boost/beast/websocket/impl/read.hpp [hunk 1]
```diff
@@ -514,8 +514,12 @@ class stream<NextLayer, deflateSupported>::read_some_op
                     }
                 }
                 BOOST_ASSERT( ! impl.rd_done );
-                if( impl.rd_remain == 0 && impl.rd_fh.fin )
+                if (impl.rd_remain == 0 && impl.rd_fh.fin) {
                     impl.rd_done = true;
+                }
+                else {
+                    cont = true;
```
Comment (ashtum) [inline include/boost/beast/websocket/impl/read.hpp:521]: `cont` is unrelated to the logic of the read operation itself; it determines whether the completion handler should be scheduled on an executor or executed directly. I suspect there might be an issue in your custom stream type, where an additional dispatch can make it work.

**Comment by pyrist** [comment]
@ashtum Thanks so much for the help. I used your example to compare what was going on with our sockets. You're right, our socket failing to call an additional dispatch is what's actually going on here, and setting cont to true just forces beast to do the socket's job for it.

**Comment by pyrist** [comment]
@ashtum Thanks so much for the help. I used your example to compare what was going on with our sockets. You're right, our socket failing to call an additional dispatch is what's actually going on here, and setting cont to true just forces beast to do the socket's job for it.

**Pull request closed without merging**